### PR TITLE
Fixing up trace collection for performance metrics

### PR DESCRIPTION
**Pull request by t0yv0**
On top of https://github.com/pulumi/pulumi-trace-tool/pull/2

Example run: https://github.com/pulumi/templates/actions/runs/2878948748

### CODE REVIEW

File: .github/workflows/performance_metrics_cron.yml [hunk 1]
```diff
@@ -98,7 +103,22 @@ jobs:
...
           echo "PULUMI_TRACING_DIR=$PWD/traces"                      >> $GITHUB_ENV
       - name: Running Performance Matrix
-        run: make test_templates
+        shell: bash
+        # Even if some Go tests fail, others may have succeded and
+        # produced perf data; continue to try to upload it.
+        continue-on-error: true
+        run: |
+          set -euo pipefail
+          cd tests && go test -v -json -count=1 -cover -timeout 6h -parallel ${{ env.TESTPARALLELISM }} . 2>&1 | tee /tmp/gotest.log | gotestfmt
```
Comment (t0yv0) [inline .github/workflows/performance_metrics_cron.yml:112]: gotestfmt makes output easier to read

File: .github/workflows/performance_metrics_cron.yml [hunk 2]
```diff
@@ -98,7 +103,22 @@ jobs:
           echo "PULUMI_TRACING_TAG_PULUMI_VERSION=$(pulumi version)" >> $GITHUB_ENV
           echo "PULUMI_TRACING_DIR=$PWD/traces"                      >> $GITHUB_ENV
       - name: Running Performance Matrix
-        run: make test_templates
+        shell: bash
+        # Even if some Go tests fail, others may have succeded and
+        # produced perf data; continue to try to upload it.
+        continue-on-error: true
```
Comment (t0yv0) [inline .github/workflows/performance_metrics_cron.yml:109]: This is key to get data even if some tests fail

File: .github/workflows/performance_metrics_cron.yml [hunk 3]
```diff
@@ -98,7 +103,22 @@ jobs:
...
+          set -euo pipefail
+          cd tests && go test -v -json -count=1 -cover -timeout 6h -parallel ${{ env.TESTPARALLELISM }} . 2>&1 | tee /tmp/gotest.log | gotestfmt
+        env:
+          PULUMI_PYTHON_CMD: python
+          TESTPARALLELISM: 3
+      - name: Upload traces GHA artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: traces
+          path: traces
```
Comment (t0yv0) [inline .github/workflows/performance_metrics_cron.yml:120]: This can help debug issues with the data

File: Makefile [hunk 4]
```diff
@@ -1,8 +1,15 @@
...
+TESTFLAGS       := -v -count=1 -cover -timeout 6h -parallel $(TESTPARALLELISM)
 
 test_templates::
-	cd tests && go test -v -count=1 -cover -timeout 6h -parallel $(TESTPARALLELISM) .
+	cd tests && go test $(TESTFLAGS)
+
+# Run a test of a single template.
+# Example: make test_template.typescript
+# This will run a test corresponding to the typescript template
+test_template.%:
```
Comment (t0yv0) [inline Makefile:11]: This helped me debug locally

File: tests/go.mod [hunk 5]
```diff
@@ -3,7 +3,7 @@ module github.com/pulumi/templates/v2
 go 1.17
 
 require (
-	github.com/pulumi/pulumi-trace-tool v0.0.0-20210629202931-10ae87ca9439
+	github.com/pulumi/pulumi-trace-tool v0.0.0-20220817225834-706e9fc5373d
```
Comment (t0yv0) [inline tests/go.mod:6]: One more round of updates needed here after upstream PR merges.

File: tests/testing_utils_test.go [hunk 6]
```diff
@@ -17,8 +17,10 @@ func runWithTimeout(
 		timeoutEvent := time.After(timeout)
 		done := make(chan bool)
 		go func() {
+			defer func() {
```
Comment (t0yv0) [inline tests/testing_utils_test.go:20]: Apparently run(t) panics occasionally making tests hang; so defer is the right thing to do.

File: tests/template_test.go [hunk 7]
```diff
@@ -87,17 +87,16 @@ func TestTemplates(t *testing.T) {
...
-			t.Logf("Skipping template test %s", templateName)
-			continue
-		}
-
-		e := ptesting.NewEnvironment(t)
-		t.Cleanup(func() { deleteIfNotFailed(e) })
-
 		runWithTimeout(t, testTimeout, templateName, parallel, func(t *testing.T) {
+			if isBlackListedTest(templateName, blackListed) {
+				t.Skip("Skipping per BLACK_LISTED_TESTS")
```
Comment (t0yv0) [inline tests/template_test.go:92]: This form of skipping is better displayed via gotestfmt

File: .github/workflows/performance_metrics_cron.yml [hunk 8]
```diff
@@ -98,7 +103,22 @@ jobs:
           echo "PULUMI_TRACING_TAG_PULUMI_VERSION=$(pulumi version)" >> $GITHUB_ENV
           echo "PULUMI_TRACING_DIR=$PWD/traces"                      >> $GITHUB_ENV
       - name: Running Performance Matrix
-        run: make test_templates
+        shell: bash
```
Comment (RobbieMcKinstry) [inline .github/workflows/performance_metrics_cron.yml:106]: I've not seen this specified before. Is Bash not the default?

File: .github/workflows/performance_metrics_cron.yml [hunk 9]
```diff
@@ -98,7 +103,22 @@ jobs:
...
+        # Even if some Go tests fail, others may have succeded and
+        # produced perf data; continue to try to upload it.
+        continue-on-error: true
+        run: |
+          set -euo pipefail
+          cd tests && go test -v -json -count=1 -cover -timeout 6h -parallel ${{ env.TESTPARALLELISM }} . 2>&1 | tee /tmp/gotest.log | gotestfmt
+        env:
+          PULUMI_PYTHON_CMD: python
+          TESTPARALLELISM: 3
```
Comment (RobbieMcKinstry) [inline .github/workflows/performance_metrics_cron.yml:115]: You could probably pull this to the top of the file if you wanted. Optional

File: .github/workflows/performance_metrics_cron.yml (hunk 3 above)
Comment (RobbieMcKinstry) [reply .github/workflows/performance_metrics_cron.yml:120]: This is super smart!

File: tests/testing_utils_test.go [hunk 10]
```diff
@@ -17,8 +17,10 @@ func runWithTimeout(
...
+				done <- true
+			}()
```
Comment (RobbieMcKinstry) [inline tests/testing_utils_test.go:22]: I thought you had to insert a `recover` statement here, but upon reading the [defer/panic/recover](https://go.dev/blog/defer-panic-and-recover) post, I see that I was wrong. 

```golang
			defer func() {
			        if r := recover(); r != nil {
			            t.Error("run failed.")
                                    done <- true
                                }				
			}()
```

I believe your code is correct as is.

**Review by RobbieMcKinstry — APPROVED** [review]
Nice! Very cool, learning a lot :)

File: .github/workflows/performance_metrics_cron.yml (hunk 8 above)
Comment (t0yv0) [reply .github/workflows/performance_metrics_cron.yml:106]: I think not - the other code had it but I'm not sure.

File: .github/workflows/performance_metrics_cron.yml (hunk 9 above)
Comment (t0yv0) [reply .github/workflows/performance_metrics_cron.yml:115]: Copying verbatim from the cron yml.

File: tests/testing_utils_test.go (hunk 10 above)
Comment (t0yv0) [reply tests/testing_utils_test.go:22]: I was empirically running into hangs that were fixed by this. Definitely learned something about Go also - defer is taking panics into account as well as early (error) returns.

**Pull request merged**